### PR TITLE
Added ability to fetch and pull coinbase orderbooks

### PR DESCRIPTION
**Pull request by stevenewald**
Code lacks significant customizability and binance data but will be added in future PRs

### CODE REVIEW

**Comment by stevenewald** [comment]
will fix ci warnings

**Comment by egelja** [comment]
Don’t worry about CI right now, it’s not working on main either.

File: src/main.cpp [hunk 1]
```diff
@@ -78,23 +80,43 @@ main(int argc, const char** argv)
...
-        // Send some data
-        std::vector<uint8_t> new_data(10000, 'a');
-        size_t sent = conn->send(std::move(new_data));
-
-        log_i(main, "Send {} bytes", sent);
+        if (string_data == "deadbeef") {
+            std::string str =
+                R"({"type":"subscribe","channels":[{"name":"level2","product_ids":["ETH-USD"]}]})";
+            std::vector<uint8_t> bytes(str.begin(), str.end());
+            conn->send(bytes);
```
Comment (egelja) [inline src/main.cpp:110]:
```suggestion
            conn->send(std::move(bytes));
```

File: src/main.cpp [hunk 2]
```diff
@@ -78,23 +80,43 @@ main(int argc, const char** argv)
     log_build_info();
     log_i(libcurl, "{}", curl_version());
 
+    redisContext* c = redisConnect("127.0.0.1", 6379);
+    if (c == NULL || c->err) {
```
Comment (egelja) [inline src/main.cpp:84]:
```suggestion
    if (c == nullptr || c->err) {
```

File: src/storage/processing.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+void
+OrderbookProcessor::process_incoming_data(const std::string& string_data)
+{
+    std::variant<ObSnapshot, Update> data{};
+    auto s = glz::read_json(data, string_data);
+
+    if (s) {
+        log_e(main, "Error parsing data: {}", glz::format_error(s, string_data));
+        return;
+    }
```
Comment (egelja) [inline src/storage/processing.cpp:19]:
```suggestion
    auto err = glz::read_json(data, string_data);

    if (err) {
        log_e(main, "Error parsing data: {}", glz::format_error(err, string_data));
        return;
    }
```

File: src/storage/processing.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    else {
+        const auto& latestSnapshot = std::get<ObSnapshot>(data);
+        process_incoming_snapshot(latestSnapshot);
+        ob_to_redis(latestSnapshot.product_id);
+    }
+}
+
+void
+OrderbookProcessor::ob_to_redis(std::string product_id)
```
Comment (egelja) [inline src/storage/processing.cpp:34]:
```suggestion
OrderbookProcessor::ob_to_redis(const std::string& product_id)
```

File: src/storage/processing.cpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+{
+    std::vector<const char*> argv;
+    argv.reserve(
+        2 * table.size() + 2
+    );
+
+    argv.push_back("HMSET");
+    argv.push_back(map_id.c_str());
+
+    std::vector<std::string> kvPairs;
```
Comment (egelja) [inline src/storage/processing.cpp:108]: What is `kvPairs` used for? I fell like you could just push to `argv` here.

File: src/storage/processing.hpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    std::string time;
+    std::string type = "l2update";
+    std::string product_id;
+    std::vector<std::tuple<std::string, std::string, std::string>> changes;
+};
+
+class OrderbookProcessor {
+private:
+    std::unordered_map<std::string, ProductTracker> Orderbook;
+    redisContext* redis;
```
Comment (egelja) [inline src/storage/processing.hpp:28]:
```suggestion
    std::unordered_map<std::string, ProductTracker> orderbook_;
    redisContext* redis_;
```

File: src/storage/processing.hpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+    void process_incoming_data(const std::string& string_data);
+
+private:
+    void process_incoming_snapshot(const ObSnapshot& newOb);
+    void process_incoming_update(const Update& newUpdate);
+    void ob_to_redis(std::string product_id);
+    void map_to_redis(
+        const std::unordered_map<double, double>& table, const std::string& map_id
+    );
```
Comment (egelja) [inline src/storage/processing.hpp:41]:
```suggestion
    void process_incoming_snapshot_(const ObSnapshot& newOb);
    void process_incoming_update_(const Update& newUpdate);
    void ob_to_redis_(const std::string& product_id);
    void map_to_redis_(
        const std::unordered_map<double, double>& table, const std::string& map_id
    );
```

File: src/storage/processing.hpp [hunk 8]
```diff
@@ -0,0 +1,77 @@
+#include "common.hpp"
+#include "product_tracker.hpp"
+
+#include <glaze/glaze.hpp>
+#include <hiredis/hiredis.h>
+
+namespace raccoon {
+namespace storage {
+
+struct ObSnapshot {
```
Comment (egelja) [inline src/storage/processing.hpp:10]:
```suggestion
struct OrderbookSnapshot {
```

File: src/storage/product_tracker.hpp [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#include "common.hpp"
+
+namespace raccoon {
+namespace storage {
+struct ProductTracker {
+    std::unordered_map<double, double> bids;
+    std::unordered_map<double, double> asks;
+};
```
Comment (egelja) [inline src/storage/product_tracker.hpp:8]:
```suggestion
struct product_tracker {
    std::unordered_map<double, double> bids;
    std::unordered_map<double, double> asks;
};
```

**Review by egelja — CHANGES_REQUESTED** [review]
I like the use of `std::variant`. Code looks great overall.

One little thing: can you convert everything but classes to snake_case?

**Comment by echavemann** [comment]
nice work! +1 on ninos. not super on edge about this code for now - we will be revisitng this many times in the future.

**Comment by egelja** [comment]
> not super on edge about this code for now - we will be revisitng this many times in the future.

agreed

File: src/storage/processing.cpp (hunk 5 above)
Comment (stevenewald) [reply src/storage/processing.cpp:108]: its necessary due to the lifetimes of the strings

**Pull request merged**